### ISSUE TITLE
[BUG]: java.lang errors causes CosmosDB query thread hang due to Reactor library issue 

### ISSUE THREAD

**Issue by wangmiao1981** [Azure/azure-sdk-for-java#28707]
**Describe the bug**
When we make CosmosDB query using SDK v 4.28 and query load exceeds RU limit, the query thread hangs forever. 

***Exception or Stack Trace***
exception: { [-]
     exception_class: reactor.core.Exceptions.ReactiveException
     exception_message: java.lang.InterruptedException
     stacktrace: reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
    at reactor.core.Exceptions.propagate(Exceptions.java:392)
    at reactor.core.publisher.BlockingIterable$SubscriberIterator.hasNext(BlockingIterable.java:184)
    at 
......

**To Reproduce**
https://github.com/kushagraThapar/cosmos-java-sdk-testing/blob/master/src/main/java/com/example/common/QueryThrottlingHangIssue.java

***Code Snippet***
Please refer to the git repo. 

**Expected behavior**
The query thread should not be hang by the underlying Reactor Lib issue. A timeout config to the query is an acceptable solution. 

**Screenshots**

**Setup (please complete the following information):**
User Agent: azsdk-java-cosmos/4.28.0 Linux/5.10.75-flatcar JRE/1.8.0_202
Pod details:
4 CPU
4 GB Memory
Multiple instances running

**Additional context**
Similar issue of other iterators in Reactor Lib has been fixed, PR: https://github.com/reactor/reactor-core/pull/2766

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ X] Bug Description Added
- [ X] Repro Steps Added
- [ X] Setup information Added

**Comment by kushagraThapar** [comment]
@alzimmermsft  / @anuchandy  / @srnagar - customer is facing this issue - https://github.com/Azure/azure-sdk-for-java/issues/21693 and since there is no way to come out of it from the reactor-core hang, I am wondering if we can do on our `PagedIterable` / `PagedFlux` APIs across Azure to add a timeout or to expose a timeout API, which can timeout from our end if the reactor-core thread is stuck. 

cc @xinlian12 / @FabianMeiswinkel

**Comment by kushagraThapar** [comment]
@joshfree - please add any more thoughts you / your team may have. This needs immediate attention, thanks!

**Comment by wangmiao1981** [comment]
@kushagraThapar  Thanks for your comments!

This issue causes severe service stability problem when query is throttled by CosmosDB backend service. Please prioritize the fix. If you need any more context on the severity of the issue, we can set up a call to discuss. In addition, this issue impacts all CosmosDB SDK users.

**Comment by joshfree** [comment]
Hi folks, if this needs immediate attention please follow the instructions here to file a support ticket: https://github.com/Azure/azure-sdk-for-java/blob/main/SUPPORT.md#azure-support-tickets

@anuchandy can you please take a look at this GitHub issue and sync up with @kushagraThapar on details?

**Comment by wangmiao1981** [comment]
hi @joshfree  thanks for your input!

We logged a ServA, CosmosDB outage ticket to MSFT, which was transferred to @kushagraThapar  who helped identify the hang issue. This is the first time that I used a git to log a Production blocker. 

Can you let me know if I escalate it to MSFT CosmosDB Engineering leadership, it will expedite the fix process? 

Sorry for not knowing the git support process. But we really need to get this one fixed asap. Now, the only workaround is increasing RU.

Thanks!

**Comment by kushagraThapar** [comment]
@wangmiao1981 - as discussed offline, moving to the latest SDK should fix the issue. On the above discussion, we want to focus on the long term solution for any java.lang.error which can happen in reactor library. We are discussing it internally on an appropriate solution for this, meanwhile, following this will help you unblock from any hangs as I mentioned in the email as well: 

1. Implement `Schedulers.onHandleError()`​ API - Example usage is here - [cosmos-java-sdk-testing/FluxHangIssue.java at master · kushagraThapar/cosmos-java-sdk-testing (github.com)](https://github.com/kushagraThapar/cosmos-java-sdk-testing/blob/master/src/main/java/com/example/common/FluxHangIssue.java#L19)
2. Here are few JVM options that you can add to your application when starting the JVM process to avoid hang issues - [Recommendations for recovery upon Out Of Memory error - Knowledge Base (neo4j.com)](https://neo4j.com/developer/kb/recommendations-for-recovery-upon-out-of-memory-error/)

**Comment by wangmiao1981** [comment]
@kushagraThapar  I am a kind of disagreeing on the conclusion of the solution. Yes, we upgraded it to the latest SDK version as you suggested. However, the reason that I disagree with the solution is because we still didn't find out the clue/evidence that the issue is caused by OOM or any java.lang error.

In addition, the new SDK based on your description resolved a memory leak bug, which should be related to OOM issue. However, as you mentioned here, any other java.lang errors could hit the issue. 

Therefore, the conclusion should be that the SDK or the Core Utility provides SDK user a way to timeout/automatically exiting the thread hang.

**Comment by wangmiao1981** [comment]
@kushagraThapar  to your code in https://github.com/kushagraThapar/cosmos-java-sdk-testing/blob/master/src/main/java/com/example/common/FluxHangIssue.java#L19, there is no evidence that the hang was caused by java.lang error. We did another long search, since we hit the issue again on May 6, 7 AM PST. There is no any java.lang error. The thread is recovered only when our management thread cancels it, which is proved by the stack trace in log. 

Therefore, my specific ask in this ticket is to unblock us whenever the thread gets hang by java.lang error or any unknown error.

As we discussed on the call, if you provide a wrapper with Timeout parameter for the query call or the reactor lib call, it can guarantee that the thread will exit.

Another data point, when we were on 2.x SDK version, we did see same throttling. But since the thread was not hang, we had no problem to handle the load with application retries. However, after upgrading to 4.28, there was no such option and thus we got hang for too long time which built big lag in the pipeline. 

To sum up, **Not hanging the caller thread** is the key requirement.

**Comment by kushagraThapar** [comment]
Closing this, as this has been fixed https://github.com/Azure/azure-sdk-for-java/pull/28343
